Source/Foundation/bsfUtility/Utility/BsDynLibManager.h

```diff
@@ -5,37 +5,70 @@
 #include "Prerequisites/BsPrerequisitesUtil.h"
 #include "Utility/BsModule.h"
 
-namespace bs 
+#include "Utility/BsDynLib.h"
+
+namespace bs
 {
 	/** @addtogroup General
 	 *  @{
 	 */
 
-    /**
-	 * This manager keeps a track of all the open dynamic-loading libraries, opens them and returns references to 
+	/**
+	 * This manager keeps a track of all the open dynamic-loading libraries, opens them and returns references to
 	 * already-open libraries.
-	 *			
+	 *
 	 * @note	Not thread safe.
-     */
-    class BS_UTILITY_EXPORT DynLibManager : public Module<DynLibManager>
-    {
-    public:
-        DynLibManager() = default;
-        virtual ~DynLibManager();
-
-        /**
-         * Loads the given file as a dynamic library.
-         *
-         * @param[in]	name	The name of the library. The extension can be omitted.
-         */
-        DynLib* load(const String& name);
+	 */
+	class BS_UTILITY_EXPORT DynLibManager : public Module<DynLibManager>
+	{
+	public:
+		DynLibManager() = default;
+		virtual ~DynLibManager() = default;
+
+		/**
+		 * Loads the given file as a dynamic library.
+		 *
+		 * @param[in]	name	The name of the library. The extension can be omitted.
+		 */
+		DynLib* load(String name);
 
 		/** Unloads the given library. */
 		void unload(DynLib* lib);
 
 	protected:
-		Map<String, DynLib*> mLoadedLibraries;
-    };
+		struct DynLibDeleter final
+		{
+			void operator()(DynLib* lib) const
+			{
+				lib->unload();
```

Comment:
Does using UPtr seem good to you? Or would you also rather not do that?

I wonder if perhaps, instead of using UPtr, we should use SPtr inside of DynLibManager, and std::weak_ptr (WPtr?) for the returned-type ? Not 100% sure if that would preserve the semantics, as I think weak_ptr can be upgraded to strong_ptr. But as long as the allocator used to allocate the DynLib* stays alive longer than the SPtr does, I don't think it'll hurt anything.

Reply:
I don't like UPtrs because of having to create custom deleters. I'd prefer it over SPtr because of the potential upgrade to strong ptr you mention, it just sounds dirty even if it technically couldn't happen :)

In a simple case like this I'd personally still go with `bs_new`/`bs_delete`. But I'm not against UPtr if you think the extra safety is important.

Reply:
Actually using SPtr for storage, but returning a raw pointer instead of a weak one, might be the best case?